R/bundlePackage.R

```diff
@@ -24,136 +25,66 @@ bundlePackages <- function(appDir,
-  for (i in seq_len(nrow(deps))) {
-    name <- deps[i, "Package"]
-
-    # get package info
-    info <- as.list(deps[i, c("Source", "Repository")])
-
-    # include github package info
-    info <- c(info, as.list(deps[i, grep("Github", colnames(deps), perl = TRUE, value = TRUE)]))
-
-    # get package description; note that we need to remove the
-    # packageDescription S3 class from the object or jsonlite will refuse to
-    # serialize it when building the manifest JSON
+  deps$description <- lapply(deps$Package, function(nm) {
+    # Remove packageDescription S3 class so jsonlite can serialize
     # TODO: should we get description from packrat/desc folder?
-    info$description <- suppressWarnings(unclass(utils::packageDescription(name)))
-
-    # TODO(HW): should this code be removed? I don't see how it is ever reached
-    # if description is NA, application dependency may not be installed
-    if (is.na(info$description[1])) {
-      errorMessages <- c(
-        errorMessages,
-        paste0(
-          capitalize(assetTypeName), " depends on package \"", name, "\"",
-          "but it is not installed. Please resolve before continuing."
-        )
-      )
-      next
-    }
-
-    # validate package source (returns a message if there is a problem)
-    packageMessages <- c(packageMessages, validatePackageSource(deps[i, ]))
+    unclass(utils::packageDescription(nm))
+  })
 
-    # good to go
-    packages[[name]] <- info
-  }
+  github_cols <- grep("Github", colnames(deps), perl = TRUE, value = TRUE)
+  packages <- deps[c("Source", "Repository", github_cols, "description")]
+  packages_list <- lapply(seq_len(nrow(packages)), function(i) {
+    as.list(packages[i, , drop = FALSE])
+  })
+  names(packages_list) <- deps$Package
+  packages_list
+}
 
-  if (length(packageMessages)) {
-    # Advice to help resolve installed packages that are not available using the
-    # current set of configured repositories. Each package with a missing
-    # repository has already been printed (see snapshotDependencizes).
-    #
-    # This situation used to trigger an error (halting deployment), but was
-    # softened because:
-    #   * CRAN-archived packages are not visible to our available.packages
-    #     scanning.
-    #   * Source-installed packages may be available after a manual server-side
-    #     installation.
-    #
-    # That said, an incorrectly configured "repos" option is almost always the
-    # cause.
-    packageMessages <- c(
-      packageMessages,
-      paste0(
-        "Unable to determine the source location for some packages. ",
-        "Packages should be installed from a package repository like ",
-        "CRAN or a version control system. Check that ",
-        "options('repos') refers to a package repository containing ",
-        "the needed package versions."
-      )
+checkBundlePackages <- function(deps, call = caller_env()) {
+  not_installed <- !vapply(deps$Package, is_installed, logical(1))
```

Comment:
This logic still isn't quite right because these errors will occur before `deployApp()`  has called `saveDeployment()`, so if this fails and you're publishing with the IDE, you'll need to click through all of the dialogs again. Not yet sure how to resolve this.